File: frontend/static/js/encryption/symmetric/playfair_lib.js
class PlayfairLib {
    constructor() {
        this.alphabet = "ABCDEFGHIKLMNOPQRSTUVWXYZ";
    }

    prepareKey(key) {
        let keyStr = (new TextDecoder().decode(key)).toUpperCase().replace(/J/g, "I");
        let res = "";
        for (let char of (keyStr + this.alphabet)) {
            if (/[A-Z]/.test(char) && !res.includes(char)) res += char;
        }
        return res;
    }

    findPos(matrix, char) {
        for (let r = 0; r < 5; r++) {
            let c = matrix[r].indexOf(char);
            if (c !== -1) return [r, c];
        }
        return [0, 0];
    }

    async encrypt(plaintext, key) {
        const keyStr = this.prepareKey(key);
        const matrix = Array.from({length: 5}, (_, i) => keyStr.slice(i*5, i*5+5));
        let text = plaintext.toUpperCase().replace(/J/g, "I").replace(/\s/g, "");
        let prepared = "";
        for (let i = 0; i < text.length; i += 2) {
            let a = text[i];
            let b = (i + 1 < text.length) ? text[i+1] : 'X';
            if (a === b) { prepared += a + 'X'; i--; }
            else { prepared += a + b; }
        }
        if (prepared.length % 2 !== 0) prepared += 'X';

        let res = "";
        for (let i = 0; i < prepared.length; i += 2) {
            let [r1, c1] = this.findPos(matrix, prepared[i]);
            let [r2, c2] = this.findPos(matrix, prepared[i+1]);
            if (r1 === r2) res += matrix[r1][(c1+1)%5] + matrix[r2][(c2+1)%5];
            else if (c1 === c2) res += matrix[(r1+1)%5][c1] + matrix[(r2+1)%5][c2];
            else res += matrix[r1][c2] + matrix[r2][c1];
        }
        return btoa(res);
    }

    async decrypt(ciphertext, key) {
        const keyStr = this.prepareKey(key);
        const matrix = Array.from({length: 5}, (_, i) => keyStr.slice(i*5, i*5+5));
        const data = atob(ciphertext);
        let res = "";
        for (let i = 0; i < data.length; i += 2) {
            let [r1, c1] = this.findPos(matrix, data[i]);
            let [r2, c2] = this.findPos(matrix, data[i+1]);
            if (r1 === r2) res += matrix[r1][(c1+4)%5] + matrix[r2][(c2+4)%5];
            else if (c1 === c2) res += matrix[(r1+4)%5][c1] + matrix[(r2+4)%5][c2];
            else res += matrix[r1][c2] + matrix[r2][c1];
        }
        return res;
    }

    async generateKey() { return new TextEncoder().encode("PLAYFAIRKEY"); }
}

File: backend/encryption/symmetric/playfair_lib.py
import base64
from .base import SymmetricEncryption

class PlayfairLib(SymmetricEncryption):
    def __init__(self):
        self.alphabet = "ABCDEFGHIKLMNOPQRSTUVWXYZ" # J dışarıda

    def _prepare_key(self, key_bytes):
        key_str = key_bytes.decode('utf-8', errors='ignore').upper().replace('J', 'I')
        res = ""
        for char in key_str + self.alphabet:
            if char.isalpha() and char not in res:
                res += char
        return res

    def _get_matrix(self, key_str):
        return [key_str[i:i+5] for i in range(0, 25, 5)]

    def _find_pos(self, matrix, char):
        for r, row in enumerate(matrix):
            if char in row: return r, row.find(char)
        return 0, 0

    def encrypt(self, plaintext, key):
        key_str = self._prepare_key(key)
        matrix = self._get_matrix(key_str)
        text = plaintext.upper().replace('J', 'I').replace(" ", "")
        
        # İkilere ayır ve aynı harf ise araya X koy
        prepared = ""
        i = 0
        while i < len(text):
            a = text[i]
            b = text[i+1] if i+1 < len(text) else 'X'
            if a == b:
                prepared += a + 'X'
                i += 1
            else:
                prepared += a + b
                i += 2
        
        res = ""
        for i in range(0, len(prepared), 2):
            r1, c1 = self._find_pos(matrix, prepared[i])
            r2, c2 = self._find_pos(matrix, prepared[i+1])
            if r1 == r2: # Aynı satır
                res += matrix[r1][(c1+1)%5] + matrix[r2][(c2+1)%5]
            elif c1 == c2: # Aynı sütun
                res += matrix[(r1+1)%5][c1] + matrix[(r2+1)%5][c2]
            else: # Dikdörtgen kuralı
                res += matrix[r1][c2] + matrix[r2][c1]
        
        return base64.b64encode(res.encode('utf-8')).decode('utf-8')

    def decrypt(self, ciphertext, key):
        data = base64.b64decode(ciphertext).decode('utf-8')
        key_str = self._prepare_key(key)
        matrix = self._get_matrix(key_str)
        res = ""
        for i in range(0, len(data), 2):
            r1, c1 = self._find_pos(matrix, data[i])
            r2, c2 = self._find_pos(matrix, data[i+1])
            if r1 == r2:
                res += matrix[r1][(c1-1)%5] + matrix[r2][(c2-1)%5]
            elif c1 == c2:
                res += matrix[(r1-1)%5][c1] + matrix[(r2-1)%5][c2]
            else:
                res += matrix[r1][c2] + matrix[r2][c1]
        return res

    def generate_key(self, key_size=None):
        return b"PLAYFAIRKEY"